worker_tasks/categorize.py

```diff
@@ -46,109 +80,349 @@
+        for ambiguous_ngram in ambiguous_ngrams:
+
+            max_relatedness_score = MINIMUM_RELATEDNESS_SCORE
+            best_article = None
+
+            for article in self._memoized_related_articles[ambiguous_ngram]:
+
+                article_relatedness_score = self._get_relatedness_score(self._memoized_article_links[article], context_links)
+
+                if article_relatedness_score > max_relatedness_score:
+                    max_relatedness_score = article_relatedness_score
+                    best_article = article
+
+            if best_article:
+                assigned_articles.append(best_article)
+
+        return assigned_articles
+
+
+    def _build_semantic_graph(self, articles):
         """
-        Choose the two highest-weighted categories that don't share
-        a common word. For example, "Just" and "Just Works" should
-        not be selected as the two best.
+        Create a category graph where each vertex is a Wikipedia article
+        and the edges are weighted by relatedness scores (number of overlapping
+        links) between each pair.
         """
-        cat1, cat2 = "", ""
-        for category, weights in categories.most_common():
-            category = category.title()
-            if cat1 in cat2:
-                cat1 = category
-            elif cat2 in cat1:
-                cat2 = category
-            else:
-                break
 
-        return cat1, cat2
+        logger.debug("Building semantic graph...")
 
+        edges = []
+        vertex_attrs = {"name": []}
+        edge_attrs = {"weight": []}
 
-    def is_from_titles(self, word):
-        # Case-Insensitive Comparison
-        # since titles are uppercase and body text lowercase
-        return word.title() in self.title_categories
+        for i, article_1 in enumerate(articles):
+            vertex_attrs["name"].append(article_1)
 
+            for j, article_2 in enumerate(articles):
+                if i == j:
+                    continue
 
-    def is_bigram(self, word):
-        return " " in word
+                if (article_1, article_2) in self._memoized_semantic_relatedness_scores:
+                    relatedness_score = self._memoized_semantic_relatedness_scores[(article_1, article_2)]
+                else:
+                    relatedness_score = self._get_relatedness_score(self._memoized_article_links[article_1], self._memoized_article_links[article_2])
+                    self._memoized_semantic_relatedness_scores[(article_1, article_2)] = relatedness_score
+                    self._memoized_semantic_relatedness_scores[(article_2, article_1)] = relatedness_score
 
+                if relatedness_score:
+                    edges.append([i, j])
+                    edge_attrs["weight"].append(relatedness_score)
 
-    def is_valid_category(self, category):
+        return igraph.Graph(edges=edges, vertex_attrs=vertex_attrs, edge_attrs=edge_attrs)
```

Comment:
It's directed, i and j will eventually swap since they both loop through the full articles array. The only one it skips is i == j (where the edge goes to itself).